src/persist-client/src/write.rs

```diff
@@ -90,6 +90,18 @@ where
         &self.upper
     }
 
+    /// Fetches and returns the shard-global `upper`.
+    ///
+    /// This requires fetching the latest state from consensus and is therefore a potentially
```

Comment:
this should document that it's not linearized with other write operations (we could also offer a linearized version with a no-op state change if that ever becomes necessary). maybe even call this `fetch_recent_upper` to really drive it home?